### PR TITLE
Handle goto definition in relative imports

### LINKED ISSUE #1022: Go to definition (F12 or Ctrl+LMB) fails when clicking on a relative import
@Trolldemorted commented on [Mon Apr 29 2019](https://github.com/Microsoft/vscode-python/issues/5507)

## Environment data

- VS Code version: 1.33.1
- Extension version (available under the Extensions sidebar): 2019.4.11987
- OS and version: W10 N Professional 1809 (Build 17763.437)
- Python version (& distribution if applicable, e.g. Anaconda): Python 3.7.3 64-bit
- Type of virtual environment used (N/A | venv | virtualenv | conda | ...): happens with and without virtualenv
- Relevant/affected Python packages and their versions: All with relative imports

## Expected behaviour

Go to definition works

## Actual behaviour

Go to definition fails ("No definition found for X"). Additionaly, showing definitions (from type hints) of members of these modules is not reliable.

## Steps to reproduce:
1. execute `code .` in an empty folder
2. create a new file (e.g. test.py)
3. type `import requests`
4. put cursor on "requests", press F12, requests' __init__.py opens as expected
5. put cursor on `.exceptions` in the line `from .exceptions import RequestsDependencyWarning`, press F12

grafik

## Logs
Output for `Python` in the `Output` panel (`View`→`Output`, change the drop-down the upper-right of the `Output` panel to `Python`)

```
Starting Microsoft Python language server.
[Info  - 16:59:50] GetCurrentSearchPaths C:\Users\Benni\AppData\Local\Programs\Python\Python37\python.exe 
[Info  - 16:59:50] Python search paths:
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37\dlls
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37\lib
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37\lib\site-packages
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37\lib\site-packages\enochecker-0.0.1-py3.7.egg
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37\lib\site-packages\psutil-5.6.1-py3.7-win-amd64.egg
[Info  - 16:59:50]     c:\users\benni\appdata\local\programs\python\python37\lib\site-packages\elastic_apm-4.2.2-py3.7-win-amd64.egg
[Info  - 16:59:50] Configuration search paths:
[Info  - 17:00:00] Microsoft Python Language Server version 0.2.63.0
[Info  - 17:00:00] Initializing for C:\Users\Benni\AppData\Local\Programs\Python\Python37\python.exe
```   

Output from `Console` under the `Developer Tools` panel (toggle Developer Tools on under `Help`; turn on source maps to make any tracebacks be useful by running `Enable source map support for extension debugging`)

```
[Extension Host] Python Extension: Cached data exists getEnvironmentVariables, extensions
console.ts:134 [Extension Host] Python Extension: Cached data exists getEnvironmentVariables, extension-output-#2
console.ts:134 [Extension Host] Python Extension: Cached data exists getEnvironmentVariables, c:\Users\Benni\vscodetest
console.ts:134 [Extension Host] Python Extension: Cached data exists getEnvironmentVariables, extension-output-#2
console.ts:134 [Extension Host] Python Extension: Cached data exists getEnvironmentVariables, c:\users\benni\appdata\local\programs\python\python37\lib\site-packages\requests\__init__.py
```

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Fixes #1022

### CODE REVIEW

File: src/LanguageServer/Impl/Sources/DefinitionSource.cs [hunk 1]
```diff
@@ -46,71 +46,146 @@ internal sealed class DefinitionSource {
...
+        private Reference TryFromVariable(string name, IDocumentAnalysis analysis, SourceLocation location, Node statement, out ILocatedMember definingMember) {
+            definingMember = null;
+
+            var m = analysis.ExpressionEvaluator.LookupNameInScopes(name, out var scope);
+            if (m == null || !(scope.Variables[name] is IVariable v)) {
+                return null;
+            }
+
+            definingMember = v;
+            if (statement is ImportStatement || statement is FromImportStatement) {
```
Comment (AlexanderSher) [inline src/LanguageServer/Impl/Sources/DefinitionSource.cs]: Do we care about to what part of the statement `location` belongs to?

File: src/LanguageServer/Test/GoToDefinitionTests.cs [hunk 2]
```diff
@@ -235,6 +235,35 @@ class A(object):
             reference.uri.Should().Be(modPath);
         }
 
+        [TestMethod, Priority(0)]
+        public async Task GotoModuleSourceFromRelativeImport() {
+            var modAPath = TestData.GetTestSpecificUri("a.py");
+            var modBPath = TestData.GetTestSpecificUri("b.py");
+
+            var root = TestData.GetTestSpecificRootUri().AbsolutePath;
+            await CreateServicesAsync(root, PythonVersions.LatestAvailable3X);
```
Comment (AlexanderSher) [inline src/LanguageServer/Test/GoToDefinitionTests.cs]: We should probably add overload to the `CreateServicesAsync` - `TestData.GetTestSpecificRootUri().AbsolutePath` is used in majority of tests.

File: src/LanguageServer/Impl/Sources/DefinitionSource.cs (hunk 1 above)
Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Sources/DefinitionSource.cs]: If it gets here it is a variable of some sort like `C` in `from A import B as C`. We do care below if we are on a variable which has parent (aka linked from another module) so we can go there.

File: src/LanguageServer/Test/GoToDefinitionTests.cs (hunk 2 above)
Comment (MikhailArkhipov) [reply src/LanguageServer/Test/GoToDefinitionTests.cs]: Sure

**Pull request merged**